feat(ourlogs): Adjust 'log' protocol per sdk feedback

k-fish opened the pull request:
### Summary
Now that we've mostly finalized the logs protocol with sdks (see [develop doc for more info](https://github.com/getsentry/sentry-docs/pull/12920)), we want to update Relay to allow the 'log' item type to be sent in this format.

SDK's will likely primarily use the `log` ItemType instead of `otel_log` since we don't need to change timestamp conventions, can send a simplified `level` (see `OurLogLevel` added in this PR). 

#### Schema changes
We've deprecated some fields in the protocol that only exist for OTEL:
- `severity_number` and `severity_text`, we're coercing these to `level`, but we're keeping the original severity text and number as attributes as OTel allows custom severity text.
- `observed_timestamp_nanos` is always set by relay regardless of what is sent because Relay is the 'collector'. We have to leave this as an attribute as well since it's being used by the existing consumer for origin timestamp.
- `timestamp_nanos` becomes `timestamp: Timestamp`
- `trace_flags`, this is unused, and the consumer doesn't even store it in the table. Will decide what to do with this later.

#### Future work
- The `ourlog_merge_otel` function can be trimmed down since we won't need to fill in deprecated fields to send the same data to the kafka consumer.  
- We may need to transform the `OurLog` protocol from json received from sdks to a generic EAP "trace items" kafka message that is essentially a couple fields (eg. traceid) + a KVMap for `attributes`.

k-fish commented on tests/integration/test_ourlogs.py:
```diff
@@ -14,7 +14,32 @@
+                        "attributes": {
+                            "boolean.attribute": {"bool_value": True},
+                            "sentry.severity_text": {"string_value": "info"},
+                        },
```
@AbhiPrasad can you double check this looks right?

AbhiPrasad replied:
Yup looks correct, I pushed up 2f8b6ddf9ca068e3a5e7674a0a72bc005725c800 to test when we get fields missing.

Dav1dde commented on tests/integration/test_ourlogs.py:
```diff
@@ -19,9 +19,7 @@ def envelope_with_sentry_logs(payload: dict) -> Envelope:
-            payload=PayloadRef(
-                bytes=json.dumps(payload).encode()
-            ),
+            payload=PayloadRef(bytes=json.dumps(payload).encode()),
```
Drive by:

```suggestion
            payload=PayloadRef(json=payload),
```

Dav1dde commented on relay-event-schema/src/protocol/ourlog.rs:
```diff
@@ -46,11 +37,54 @@ pub struct OurLog {
+    /// The trace flags of the log.
+    #[metastructure(required = false, trim = false)]
+    #[deprecated]
+    pub trace_flags: Annotated<i64>,
```
If they are deprecated, do we need to list them in the 'schema', they would be caught by the `other` below.

k-fish replied:
I forgot that other moved the values into the struct, thanks

Dav1dde commented on relay-event-schema/src/protocol/ourlog.rs:
```diff
@@ -46,11 +37,54 @@ pub struct OurLog {
     #[metastructure(pii = "true", trim = false)]
     pub attributes: Annotated<Object<AttributeValue>>,
```
Do we want to keep this Otel Like format? Could just be a self describing mapping?

```json
            "attributes": {
                "boolean.attribute": {
                    "bool_value": true
                },
```

Could just be:

```json
            "attributes": {
                "boolean.attribute": true,
```

Dav1dde replied:
@cleptric 👀

k-fish replied:
I asked for sdk folks to discuss in the develop PR, so we'll see.

cleptric replied:
https://github.com/getsentry/sentry-docs/pull/12920/files#r2012403560

k-fish commented on relay-ourlogs/src/ourlog.rs:
```diff
@@ -61,27 +93,152 @@ pub fn otel_to_sentry_log(otel_log: OtelLog) -> OurLog {
+                        }
+                    }
+                }
+            }
```
@Dav1dde what's the preferred way of flattening these option chains? `and_then`?

Dav1dde replied:
Yes unfortunately that's very verbose, `and_then` is usually the way if it is too big, you can make a helper function which returns `Option<T>` and then use the `?` for example:

```rs
fn foo(t: Option<Option<Option<i32>>>) -> Option<i32> {
    Some(t???)
}
```

k-fish replied:
Thanks, I'll take a look

Dav1dde commented on tests/integration/test_ourlogs.py:
```diff
@@ -174,32 +164,66 @@ def test_ourlog_extraction_with_sentry_logs(
     }
-    expected_with_missing_fields = {
+
+    del ourlogs[0]["received"]
```
We have a test helper which can fuzzy assert the time, `time_within{_delta}`: https://github.com/getsentry/relay/blob/bb8969728b9d3398138b7c5acd25d2da73a22f87/tests/integration/test_profile_chunks.py#L137-L149

Then you don't have to delete the time.

Dav1dde commented on tests/integration/test_ourlogs.py:
```diff
@@ -136,35 +136,25 @@ def test_ourlog_extraction_with_sentry_logs(
     ourlogs = ourlogs_consumer.get_ourlogs()
-    assert len(ourlogs) == 2
-    expected_with_all_fields = {
+    assert len(ourlogs) == 1
```
Can just `assert ourlogs = [expected]` instead of asserting length and first element.

Dav1dde commented on tests/integration/test_ourlogs.py:
```diff
@@ -174,32 +164,66 @@ def test_ourlog_extraction_with_sentry_logs(
-    assert ourlogs[1] == expected_with_missing_fields
-
-    ourlogs_consumer.assert_empty()
+    assert ourlogs[0] == expected
```
Same here, `assert ourlogs = [expected]` and you don't even have to assert the length.

Dav1dde commented on tests/integration/test_ourlogs.py:
```diff
@@ -174,32 +164,66 @@ def test_ourlog_extraction_with_sentry_logs(
+        },
     }
-
     del ourlogs[0]["received"]
```
Also here can use the `time_within_delta` helper (as it's already used for the `timestamp_nanos`)

Dav1dde commented on relay-ourlogs/src/ourlog.rs:
```diff
@@ -131,76 +131,59 @@ pub fn otel_to_sentry_log(otel_log: OtelLog) -> OurLog {
+            ))),
+        );
+        attributes.insert(
+            "sentry.severity_text".to_string(),
```
```suggestion
            "sentry.severity_text".to_owned(),
```

We generally use `to_owned` for `&str` -> `String`. It expresses the intent better and it doesn't go through the `Display` trait to do the conversion.

Dav1dde reviewed (commented):
Just some tiny nits

k-fish committed:
feat(ourlogs): Adjust 'log' protocol per sdk feedback

Now that we've mostly finalized the logs protocol with sdks (see develop doc for more info), we want to update Relay to allow the 'log' item type to be sent in this format.

Also in this PR is some shimming between the updated protocol / event schema and the kafka consumers. This shimming is temporary until the generic EAP items consumers are ready, at which point we'll have to transform the event-schema Relay accepts into the generic eap item kafka.

k-fish commented on relay-event-schema/src/protocol/ourlog.rs:
```diff
@@ -154,90 +159,319 @@ impl Empty for AttributeValue {
+                        meta.set_original_value(Some(Value::Object(object)));
+                        Annotated(None, meta)
+                    }
+                }
```
We're now accepting:

```json
{
  "myattribute": {
     "value": "foo",
     "type": "string"
  }
}
```

Instead of: 
```json
{
  "myattribute": {
     "string_value": "foo"
  }
}
```

Although we still serialize using `string_value` (for now) until the Kafka schemas are updated / replaced by EAP items.

k-fish commented on relay-event-schema/src/protocol/ourlog.rs:
```diff
@@ -154,90 +159,319 @@ impl Empty for AttributeValue {
+
+                let value = object.remove("value");
+
+                match (attribute_type.as_deref(), value) {
```
My rust power level is too low to know whether there is a better way than cloning, open to suggestions

Dav1dde commented on relay-event-schema/src/protocol/attribute.rs:
```diff
@@ -0,0 +1,17 @@
+    Int { value: Annotated<i64> },
+    Double { value: Annotated<f64> },
+    Bool { value: Annotated<bool> },
+}
```
Dead code?

k-fish replied:
Yeah, I figured we'd need to share `Attribute` but then decided it was probably out of scope for this PR (which it seems like you agree with)

Dav1dde reviewed (changes requested):
The implementation has a differing implementation for `FromValue` and `ToValue` with the implicit assumption that the only serialization happens for Kafka. But an event can flow through multiple Relay instances, this will not work. `FromValue` and `ToValue` must agree on the format. I recommend updating the consumers first.

On that note, the `Attribute` implementation is not forward compatible, but it is also very hard to implement. I'll work on this in a separate PR which you then can depend on.

k-fish committed:
Merge remote-tracking branch 'origin/master' into feat/ourlogs/adjust-log-item-type

k-fish committed:
Switch to using flattened structure in the protocol

k-fish committed:
Rm dead code

k-fish committed:
Add structs for LogKafkaMessage to fix serialization of attributes

k-fish committed:
Get processing working

cleptric commented on relay-event-schema/src/protocol/ourlog.rs:
```diff
@@ -1,243 +1,437 @@
         match self {
-            AttributeValue::StringValue(s) => Some(s),
-            _ => None,
+            Self::Bool => "bool",
```
```suggestion
            Self::Bool => "boolean",
```

cc @AbhiPrasad @Dav1dde

cleptric commented on relay-event-schema/src/protocol/ourlog.rs:
```diff
@@ -1,243 +1,437 @@
-            AttributeValue::StringValue(s) => Some(s),
-            _ => None,
+            Self::Bool => "bool",
+            Self::Int => "int",
```
```suggestion
            Self::Int => "integer",
```

cc @AbhiPrasad @Dav1dde

cleptric reviewed (commented):
Just some naming updates

k-fish committed:
Update relay-event-schema/src/protocol/ourlog.rs

Co-authored-by: Michi Hoffmann <cleptric@users.noreply.github.com>

k-fish committed:
Update relay-event-schema/src/protocol/ourlog.rs

Co-authored-by: Michi Hoffmann <cleptric@users.noreply.github.com>

k-fish committed:
Fix tests

k-fish committed:
Fix types to match spec

k-fish committed:
Fill out the rest of the types in tests and check for forward compatibility (allow other to be sent but not emit it to kafka)

k-fish committed:
Include 'other' in debug

k-fish committed:
Fix test snapshot

k-fish committed:
process attributes without clone

k-fish committed:
Again without using clone

k-fish committed:
u64 i64 for f64

cleptric reviewed (approved):
Protocol in tests looks good to me 👍

Dav1dde commented on relay-ourlogs/src/ourlog.rs:
```diff
@@ -34,54 +71,214 @@ pub fn otel_to_sentry_log(otel_log: OtelLog) -> OurLog {
+            }
+        }
+
+        if let Some(value) = ourlog_value.attribute("sentry.severity_number") {
+            if let Some(v) = value.value() {
+                ourlog_value
+                    .other
+                    .insert("severity_number".to_string(), Annotated::new(v.clone()));
+            }
+        }
+
+        if let Some(value) = ourlog_value.attribute("sentry.trace_flags") {
+            if let Some(v) = value.value() {
+                ourlog_value
+                    .other
+                    .insert("trace_flags".to_string(), Annotated::new(v.clone()));
+            }
+        }
+
+        if let Some(value) = ourlog_value.attribute("sentry.observed_timestamp_nanos") {
+            if let Some(s) = value.as_str() {
+                if let Ok(nanos) = s.parse::<u64>() {
+                    ourlog_value.other.insert(
+                        "observed_timestamp_nanos".to_string(),
+                        Annotated::new(Value::U64(nanos)),
+                    );
+                }
+            }
+        }
+
+        if let Some(value) = ourlog_value.attribute("sentry.timestamp_nanos") {
+            if let Some(s) = value.as_str() {
+                if let Ok(nanos) = s.parse::<u64>() {
+                    ourlog_value.other.insert(
+                        "timestamp_nanos".to_string(),
+                        Annotated::new(Value::U64(nanos)),
+                    );
+                }
+            }
+        }
```
Writing into the `other` field **only** works, because the remove other processor is not running on the log item, which it definitely should. We don't want to forward arbitrary data into Kafka.

If you need this **now**, it's fine, but this needs to be changed ASAP and run the processor!

k-fish replied:
I originally had these as deprecated fields 😄. I think once we do some more work and switch to snuba-items kafka we can stop doing this since everything can be / will be sent as an attribute and no longer needs to conform to the logs consumer.

k-fish replied:
Captured
refs: LOGS-49

Dav1dde commented on relay-ourlogs/src/ourlog.rs:
```diff
@@ -34,54 +71,214 @@ pub fn otel_to_sentry_log(otel_log: OtelLog) -> OurLog {
+        Some(OurLogLevel::Warn) => 13,
+        Some(OurLogLevel::Error) => 17,
+        Some(OurLogLevel::Fatal) => 21,
+        _ => 25,
```
When parsing, we default other to info, is 25 correct here or should it be rather treated as info?

k-fish replied:
You're right, actually `0` is [default](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/docs/types/severity.md) and is essentially 'unmapped'.

Dav1dde committed:
minor clean ups

Dav1dde commented on relay-server/src/services/processor/ourlog.rs:
```diff
@@ -112,13 +128,54 @@ fn scrub(
+            continue;
+        };
+
+        match (&mut inner.value.ty, &mut inner.value.value) {
```
I scrapped the value conversion string -> double and string -> integer, as discussed. Let's start strict.

Dav1dde commented on relay-server/src/services/processor/ourlog.rs:
```diff
@@ -112,13 +128,54 @@ fn scrub(
+                value_meta.add_error(ErrorKind::InvalidData);
+                value_meta.set_original_value(value.take());
+            }
+            (Annotated(None, _), _) | (_, Annotated(None, _)) => {
+                let original = attribute.value_mut().take();
+                attribute.meta_mut().add_error(ErrorKind::MissingAttribute);
+                attribute.meta_mut().set_original_value(original);
+            }
```
This part should normally be already done by the `SchemaProcessor` which we definitely should enable for logs, similar to the `RemoveOther` mentioned elsewhere.

k-fish committed:
Update instas

k-fish committed:
Update severity to default to 0, include a test

k-fish committed:
Add schema processor

k-fish committed:
Fix tests

k-fish commented on relay-server/src/services/processor/ourlog.rs:
```diff
@@ -112,13 +131,54 @@ fn scrub(
+                let original = attribute.value_mut().take();
+                attribute.meta_mut().add_error(ErrorKind::MissingAttribute);
+                attribute.meta_mut().set_original_value(original);
+            }
```
@Dav1dde I was relying on overwriting to `Unknown` here so when it got transformed into LogKafkaMessage we can omit the attributes from the map entirely, not sure if there is a better way to drop Annotated with errors before the serialization? I'm pretty sure the existing consumer will throw errors if we emit an empty object. 

It's the only thing failing at the moment on the python integration test.

Dav1dde replied:
We can do the same we do in case of `None` and set the value in the map to `null` instead, e.g. with:

```rust
                let original = attribute.value_mut().take();
                attribute.meta_mut().add_error(ErrorKind::MissingAttribute);
                attribute.meta_mut().set_original_value(original);
```

But this is turning into a bigger problem, we're working here on wire protocol how SDK and Relay to Relay communication is taking place, and on that protocol nulls need to be expected anywhere.

There is no guarantee there won't be a shape like that forwarded, in the errors product the entire pipeline understands that. This is also an important property for data transparency, users can figure out why certain data is missing they definitely sent. Just dropping information here means that is lost.

Dav1dde replied:
Updated it so the entire attribute is removed.

Dav1dde committed:
remove attributes instead of just type/value

Dav1dde reviewed (approved):
As per discussion, this should *still* not be used by SDKs, there are more changes planned to the envelope protocol, just not to the log item.

https://www.notion.so/sentry/Performance-Envelope-Containers-1d28b10e4b5d80b3b1c3e661f5eda6b4
